federation/pkg/federation-controller/service/service_helper.go

```diff
@@ -0,0 +1,219 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"time"
+	"fmt"
+
+	"k8s.io/kubernetes/federation/client/clientset_generated/federation_release_1_3"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/v1"
+	cache "k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/golang/glog"
+	"reflect"
+)
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (sc *ServiceController) clusterServiceWorker() {
+	fedClient := sc.federationClient
+	for clusterName, clusterCache := range sc.clusterCache.clientMap{
+		go func() {
+			for {
+				key, quit := clusterCache.serviceQueue.Get()
+				// update service cache
+				if quit {
+					return
+				}
+				defer clusterCache.serviceQueue.Done(key)
+				err := sc.clusterCache.syncService(key.(string), clusterName, clusterCache, sc.serviceCache, fedClient)
+				if err != nil {
+					glog.V(2).Infof("failed to sync service: %+v", err)
+					//return err
+				}
+			}
+		}()
+	}
+}
+
+// Whenever there is change on service, the federation service should be updated
+func (cc *clusterClientCache) syncService(key, clusterName string, clusterCache *clusterCache, serviceCache *serviceCache, fedClient federation_release_1_3.Interface) error {
+	// obj holds the latest service info from apiserver, return if there is no federation cache for the service
+	cachedService, ok := serviceCache.get(key)
+	if !ok {
+		// here we filtered all non-federation services
```

Comment:
In order to update the federation service ingress info, we need to watch all k8s clusters for all services, so that if a service in a k8s cluster gets ingress info, the federation service can know that and reflect the change.
If we watch the k8s cluster services, so far there is no way to distinguish if the service is from federation or not, so what can we do is to get all service info and check if the service is in cache.